### ISSUE TITLE
OAuth 2.0 (RFC 6749) violations

### ISSUE THREAD

**Issue by zacknewman** [dani-garcia/vaultwarden#4426]
Sections [10.3](https://www.rfc-editor.org/rfc/rfc6749#section-10.3) and [10.4](https://www.rfc-editor.org/rfc/rfc6749#section-10.4) of RFC 6749 requires the authorization server to use TLS when exchanging the access and refresh tokens; however Vaultwarden—which acts as both the resource and authorization servers—currently does not force TLS, meaning deployments without TLS violate the RFC.

Additionally [Section 6](https://www.rfc-editor.org/rfc/rfc6749#section-6) requires the refresh token to be uniquely linked to the client; however Vaultwarden does not have a `UNIQUE CONSTRAINT` defined on `devices.refresh_token` which means it's possible, albeit unlikely, for multiple clients to have the same refresh token.

If adherence to RFC 6749 is not a goal, I do believe a `UNIQUE CONSTRAINT` should still be defined on `devices.refresh_token` since it will ensure the token is truly unique. Furthermore [Vaultwarden already (incorrectly) assumes/hopes the refresh token is unique](https://github.com/dani-garcia/vaultwarden/blob/main/src/db/models/device.rs#L201); and with an actual `UNIQUE CONSTRAINT` defined—which will be backed by a `UNIQUE INDEX` with any real-world RDBMS I am aware of—the performance will be even better (and most importantly correct).

With this change, it would also be nice if refresh token compromise were detected as well forcing all clients associated to a user to re-authenticate via password and whatever 2-factor authentication is hopefully configured.

**Comment by BlackDex** [comment]
Thanks for the report.

The enforcement of TLS/HTTPS will not added i think for multiple reasons.
1. People could use a reverse proxy which does HTTPS termination and Vaultwarden might not know if it is use or not.
2. People who use it locally (localhost/127.0.0.1, or which every resolves to the same) only do not need SSL for a working environment.

Regarding the unique, that is probably something we should adjust. Together with some lifetime/ttl checks probably.
There are some PR's open which link with this. But those need to be tested first on what the impact would be.

**Comment by BlackDex** [comment]
In general most projects, even if it is something secure or not do not enforce TLS on the container side. They leave that to the person who implements it.

We do notify in the `admin/diagnostics` that HTTPS is not enabled though.
And we also say in several places that without TLS some stuff will not work.

Only way i would see this happen is if we generate a CA and Self-Signed cert our selfs, and use that as a default.
But that seems nasty too.

**Comment by BlackDex** [comment]
Ok, let's say we enforce it. But i want my TLS to be terminated by my reverse proxy. How would you then determine i have TLS enabled there? Or would you force me to configure TLS also on Vaultwarden?

Adding a variable to allow non TLS enforcing would not make development easier as you suggested.

What would you suggest for this scenario?

**Comment by eoprede** [comment]
@zacknewman Correct me if I am wrong, but from my understanding RFC 6749 does not require that TLS termination happens on the entity serving authorization tokens. I believe Vaultwarden installation does not violate RFC 6749 as long as it is fronted by a reverse proxy or other system that does TLS termination.
With that said, I personally run Vaultwarden in k8s. It suits environment perfectly. And the last thing I need is to have 100 different places where I need to manage my certs. I already have ingress controller that manages them all, that's enough. If somebody manages to get into my systems/network and starts sniffing packets between k8s hosts - I have a whole lot of other issues. So from my perspective - Vaultwarden should focus on what it does best (being a backend providing Bitwarden API) and let another system manage TLS.

**Comment by eoprede** [comment]
@zacknewman 
The section you quoted says `Access token credentials MUST only be transmitted using TLS` and also that `MUST be kept confidential in transit and storage`. I think it demonstrates the issues between RFC and real-life implementation. Let me illustrate with examples:
1) vaultwarden terminates TLS in rust and uses that to talk to the client. This clearly fits the definition.
2) vaultwarden terminates TLS in rust, but there's a TCP load balancer between it and the client. Since TCP load balancer does not terminate TLS connection, this also clearly fits the definition.
3) vaultwarden terminates TLS in rust, but there's an HTTP load balancer between it and the client. Now this gets complicated. Load-balancer terminates the TLS session from client and then creates another TLS session to vaultwarden. It still definitely fits the definition of "transmitting using TLS", but what about confidentiality? Technically speaking, this session can be listened to (in case load balancer is compromised, for example). So is it confidential? I don't know how it is from the RFC standpoint, but I can assure you that a huge part of Oauth implementations out there will use this method - as all your major players do have extensive use of http load balancers (or reverse proxies) in their infrastructure. I'm going to argue that this implementation still fits the RFC, otherwise that means world is just ignoring the RFC and renders it useless.
4) vaultwarden runs on a host, where it does NOT do any TLS. However, there's a reverse proxy on the same host that does handle TLS. From the perspective of the traffic path, it's identical to the example 3 - traffic is encrypted during transit (meaning when the data hits the wire). There's a small hop between local rust process and local reverse proxy process where the traffic is not encrypted, and the client does not have direct TLS session straight to the rust process. So does it confirm to the definition? I'd argue that yes - we don't call passing data from one process to another a "transmission" and it's not any different than example 3 in terms of vector of attack.
5) vaultwarden runs as a k8s pod without TLS, using ingress controller (aka reverse proxy) to terminate TLS. This k8s cluster consists of just a single node. This is pretty much identical to option 4 from the traffic perspective, you just use k8s to configure vaultwarden and reverse proxy, as opposed to doing it manually. 
6)  vaultwarden runs as a k8s pod without TLS, using ingress controller (aka reverse proxy) to terminate TLS. This k8s cluster consists of multiple nodes, but it is using Calico [WireGuard](https://docs.tigera.io/calico/latest/network-policy/encrypt-cluster-pod-traffic) to encrypt internal communication. But WireGuard is not TLS, so I guess technically speaking now we are in breach of RFC as there _may_ be a case where token hits a wire without being encrypted by TLS. But in practice, security of communication for this option are pretty much identical to option 5.
7) vaultwarden runs as a k8s pod without TLS, using ingress controller (aka reverse proxy) to terminate TLS. This k8s cluster consists of multiple nodes and is using a standard Calico network plugin. Again - it does not confirm to Oauth spec, as it's not transmitted using TLS. But in reality, given a secure data center environment, stealing traffic off a wire in a locked cage is not an easy task. I'm not saying that it can't be done, but it does require physical access to the system, in which case you may as well install a keylogger to the console cart, steal credentials and compromise the system via local admin account. So again - for all the practical purposes, this setup is not any less secure than others. And other security standards realize that. For example, PCI explicitly calls out that data has to be encrypted over public networks, leaving internal networks open to implementation. 

Hopefully these examples will illustrate that there are multiple ways of deploying vaultwarden with strict adherence to RFC specs and many of them do not require vaultwarden to worry about TLS implementation. As vaulwarden is positioned as a lightweight product, I don't see any benefits to adding unnecessary features. And there are even more ways that while technically may not adhere to the spec, in reality they are just as secure.

Oh regarding ` I think that if handling X.509 v3 certificates is "too much", then you shouldn't be hosting something so important. It is quite easy to manage this`. Managing certs for 10 systems is easy. It's easy to do that for 100 systems as well. For 100,000 systems - trust me, you don't want to manage certs for each individual system. You would either implement some kind of service mesh (like VPC Lattice), or you would do TLS termination at the end and run stuff unencrypted internally. The places I have worked at (and you would know these companies) usually use combination of these methods depending on the workload.

**Comment by eoprede** [comment]
@zacknewman 
I think we may have misunderstanding. Can you explain how `Anyone can listen to the traffic between the reverse proxy server and Vaultwarden` when we are talking about communication within the same host (meaning, the vaultwarden and reverse proxy are running on the same machine/within the same VM)? The only way this can be done is if attacker gained access to the machine/vm. But if they have access to it, they can intercept data from the processes themselves. TLS won't solve anything, as the attacker can simply look at the data TLS is about to encrypt. The moment attacker has access to the host/vm, data is fully compromised with any of the options. And if attacker has no access to the host/vm, the data cannot be intercepted in any way in its unencrypted form, as it never physically leaves that host/vm.

Basically option 3 is more secure than option 6 because traffic never leaves the physical host/vm in option 3, while in option 6 it does leave the host/vm while being encrypted by something outside of TLS. And yes, to intercept data in option 7 one either have to gain remote access to a load balancer/reverse proxy (same attack vector as many other options), or gain access to the host running vaultwarden (same attack vector as all the other options), or get access to the data center, cage within it and then install a physical device on the wire. 

` I'm curious what a company like Google` Can't talk specifically about Google, but I've seen multiple companies (that employ ex-Google engineers) run service mesh with sidecar container. Basically you deploy your application in a k8s pod, your application runs as a container that listens to HTTP requests, but all the communication has to go through the sidecar container that handles encryption. Because containers in the same pod are scheduled to run on the [same node](https://kubernetes.io/docs/concepts/containers/), this is effectively identical to options 3-4 that I outlined above. The point is - you don't want your developers wasting time re-inventing the wheel and creating custom encryption in thousands (or even millions) of microservice applications you have, instead you have your security team come up with a robust solution that can be re-used with every single application. And k8s makes it very easy to run together. If it's good for big boys (and I am willing to bet Google does it too) - it's good enough for me, when it comes to running vaultwarden.

**Comment by eoprede** [comment]
@zacknewman 
` but I still fail to see how requiring TLS on Vaultwarden's side precludes such setups`
It doesn't, but it unnecessarily complicates things IMO.
`you claim that doing so will no longer make Vaultwarden "lightweight"; but I don't like blind assertions especially ones on performance without hard data. Where is the actual data that shows TLS will have a measurable negative impact?`
This is not about performance at all. It's about effort needed to maintain. Now developers of vaultwarden have also to monitor all the potential security risks and bugs in rust related to TLS implementation. And implement them as needed. That's a lot of extra time that can be spent on much better things. Why not let other teams (that develop well known reverse proxies like haproxy, nginx, etc) focus on the security aspect? But if you feel like TLS is necessary - why should vaultwarden "outsource" its database? Wouldn't creation of a custom database solution enhance the security?
`You also didn't show how using Kubernetes makes X.509 certificates "unreasonably" burdensome. `
I run personal vaultwarden on my home k8s cluster (since I do maintain them for living, running a small one at home takes minimal effort). I have a script to renew certs with letsencrypt for all the local services I run that are internet-accessible. Once it has been written, it takes exactly 0 effort to maintain. And adding a new cert is automatic - it happens as part of k8s ingress configuration. Compare 0 effort with some effort for no security benefit - I think there are zero reasons (aka unreasonable) to put that extra effort, And also don't forget another thing - maintaining it all. Imagine Let's Encrypt stops functioning. I just fix my solution for some other certificate provider and it works for all the services I host. Otherwise, I'd have to go and figure out solution for every single service I have - that's again unnecessary amount of effort. 
`For example can you prove to me that reusing the same X.509 certificate the reverse proxy server uses to enforce TLS is "too complex"?`
So, first thing I'd have to figure out is how reverse proxy accesses container in a pod. I think that's done by internal IP. That alone makes it impossible to reuse the same cert - because cert is created for vaultwarden.mydomain.com, while internal system will be accessed at, let's say, 10.25.67.12. Which means it's going to be an invalid cert, so I may as well use a self-signed cert. But that's insecure, right? Which means I have to run my own certificate authority, and figure out how to integrate that with k8s, how to issue and revoke certs once the pod restarts (as it gets different IPs every time), how to then integrate that code with vaultwarden container (because I need to know which IP container has to generate cert, but the IP is allocated only once container is created, so I'd have to create container with dummy cert, then generate a new cert, then replace it and then restart the service). I hope you are seeing how this gets very complicated very quickly. 
Oh, and that's just my setup in my k8s cluster. Somebody else may choose to run their vaultwarden in AWS and use AWS's Load-Balancer with AWS-managed certificate in front. You can not get that cert. there's simply no way to download it. It lives only on the load balancer and nowhere else. Makes it quite difficult to reuse it.
Oh, and... I can keep coming up with examples, but hopefully you get the point - it's actually not easy at all and each solution has to be custom made for each environment to be robust.
` As time goes on, we are seeing services integrate cryptography directly even at the protocol level (e.g., HTTP/3)`
Sure, but let's not forget that many backend services still run HTTP/1, while HTTP/2 is provided by reverse proxies in front of them. Who says HTTP/3 won't be the same?
I agree that true end to end encryption is a great goal. But operating it isn't easy. I am sure big guys have enough engineering resources (maybe, though maybe not with all the layoffs) to pull it off. But nobody else will do it properly until it's made very simple. And I hope that you see it's nothing but simple.

**Comment by BlackDex** [comment]
Let's agree to disagree on this topic. 
In regards to HTTP2/HTTP3 that is not something currently supported by Rocket (at least not there version we use). But that can be solved via a reverse proxy too 🎉. 

Also whether we enabled TLS or not, webauthn will not work without it, so in that sense it's mandatory. 

What @eoprede means in regards to managing certificates. It's fine to manage a few manually, it becomes a pain in the ... if you need to do this for all separate services when you have a dynamic environment like k8s and run maybe 30 or more different service and all those services (like Vaultwarden) have different ways on how to configure where the certificate is located, how you can enable or disable TLS versions, ciphers, MAC's etc... (if at all), you can configure a reverse proxy just once in a secure way and even allow or force new http versions. And also keep it secure and up to date for all services in one go.

Also Vaultwarden will still need HTTP/1.1 for websocket connections. So that will be the minimum version until something else is going to be used.

**Comment by BlackDex** [comment]
@zacknewman , as always you are free to provide a good PR which could help do this. But as we already know, you are not going to do this. 

We will pickup the database adjustments and maybe a bit more, so thanks for reporting this specific issue and bring it to our and others attention.

**Comment by BlackDex** [comment]
Also, regarding the PR's, and especially the attachment one, I do not recall you provide any PR at all. And i would never reject a PR or if someone is offering to create one. My memory could fail me of course. If you revere to the sentence `I can attempt a diff myself if preferred`, then for me that was a bit ambiguous, and i nowhere disregard it in any specific form. I just mentioned we would take a look at it. That shouldn't have prevent anyone who is willing and able to provide a PR that would fix it in my opinion. I think that specific item is a case of miscommunication or misunderstanding.

**Comment by BlackDex** [comment]
> Back to the issue of enforcing TLS to guarantee RFC 6749 and WebAuthn conformance. The reason I was against deploying a self-signed certificate is that I don't think it makes TLS enforcement appreciably easier. Deployments that rely on TLS via Rocket are unaffected entirely since they will remain using the X.509 v3 certificate they are already using. Deployments that rely on TLS via an intermediary (e.g., reverse proxy server) benefit very little. Specifically such deployments wouldn't have to create a self-signed certificate themselves, but they would still have to configure their intermediary to trust the certificate. Is that worth it? If you think that is the only way TLS enforcement works, then fine.
> 

I don't think that. What other solutions (besides forcing certificates) do you see for this challenge?

> Of course that would mean that insecure deployments are strictly forbidden; and as you stated:
> 
> > Also whether we enabled TLS or not, webauthn will not work without it, so in that sense it's mandatory.
> 
> that seems to be acceptable. If TLS is not enforced, then I think the project should emphasize the consequences of that even if it should be "obvious" to most people. Specifically all data that is sent is in plaintext which means the encrypted vault, OAuth tokens, master password hash, and potentially other secrets like TOTP keys when a user looks at the value (in addition to the initial time it was set up). This in turn means that deployments that enable TLS later may not be any more secure since an entity could have already captured said data and use it.

Where should we emphasize this? We already provide a note in the admin/diagnostics page. We have a better warning in the web-vault then Bitwarden if someone tries to use the Crypt API without SSL, because that also does not work.


I'm searching for good solutions here, since, as i stated before, I'm also for security, but it indeed should not become an annoyance to configure. Do you maybe know other software/packaged solutions which do something like this which we could check?

**Comment by stefan0xC** [comment]
From my point of view this issue should be closed / turned into a discussion (as it's not really a technical problem with this project or something that will be fixed by a PR but seems like a difference of opinion about security vs usability).

**Comment by stefan0xC** [comment]
@zacknewman Thanks for the clarification. I still think that the issue should be closed / turned into a discussion because there's no disagreement about that first issue and you already agreed to disagree on the second issue.

**Comment by xxsuperdreamxx** [comment]
@zacknewman, RFC 6749 is a Proposed Standard. This means that no one is “forced” to follow that standard in every single detail. There is a reason why OAuth 2.1 is still in draft form and being updated.

Regarding concerns about security behind reverse proxies, there are two relevant standards you can consider:

[RFC 9209: The Proxy-Status HTTP Response Header Field]((https://www.rfc-editor.org/rfc/rfc9209.pdf) and [RFC 9440: Client-Cert HTTP Header Field](https://www.ietf.org/rfc/rfc9440.pdf) provides guidance on handling proxy-related information in HTTP responses.
[RFC 9440: Client-Cert HTTP Header Field](https://www.ietf.org/rfc/rfc9440.pdf) discusses how a reverse proxy can make the TLS client certificate available to the backend application. (you can use this for request declining from your webserver)

Many companies adopt the practice of using locally hosted backends that are not exposed to the internet directly. These backends operate behind reverse proxies or load balancers and to SSL Offloading [Article 1](https://venafi.com/blog/what-tlsssl-offloading/) and [Article 2](https://www.miniorange.com/reverse-proxy/ssl-encryption). In such cases, VaultWarden is not required to enforce SSL. Instead, you can configure your DNS server (e.g., Cloudflare) to allow communication only to servers with the same root CA certificate.

Lastly, RFC 6749 is an older standard. We now have [RFC 8996: Deprecating TLS 1.0 and TLS 1.1](https://datatracker.ietf.org/doc/html/rfc8996), which addresses TLS security. It has been approved by the IETF as a Best Current Practice. TLS 1.2, mentioned in RFC 6749, never received approval, and we have since moved to [TLS 1.3 as the new standard](https://datatracker.ietf.org/doc/html/rfc8446).

Keep in mind that these papers primarily apply to widespread public production environments. VaultWarden, like Bitwarden, is considered an enterprise product for internal company use and for companies with solid IT knowledge. If you host it yourself, you can take control of its security

**Comment by Muska-Ami** [comment]
+1 for "this issue should be closed / turned into a discussion"

**Comment by stettler** [comment]
> Second, binding Vaultwarden to the loopback interface is an exceptional case since we are then talking about an environment that has only one client device. 

imho that's a common case: If you put it behind a reverse proxy hosted on the same server, there is no reason to expose it by binding it to anything else than the loopback interface.
